src/modelframe.jl

```diff
@@ -131,31 +131,67 @@ function evalcontrasts(df::AbstractDataFrame, contrasts::Dict = Dict())
     return evaledContrasts
 end
 
+
+## copied from DataFrames:
+function _nonmissing!(res, col)
+    # workaround until JuliaLang/julia#21256 is fixed
+    eltype(col) >: Missing || return
+    
+    @inbounds for (i, el) in enumerate(col)
+        res[i] &= !ismissing(el)
+    end
+end
+
+function _nonmissing!(res, col::CategoricalArray{>: Missing})
+    for (i, el) in enumerate(col.refs)
+        res[i] &= el > 0
+    end
+end
+
+# handle some inconsistencies between how named tuples are constructed
+if isdefined(Core, :NamedTuple)
+    _select(d::Data.Table, cols::NTuple{N,Symbol} where N) = NamedTuple{cols}(d)
+    _filter(d::T, rows) where T<:Data.Table = T(([col[rows] for col in d]...))
+else
+    _select(d::Data.Table, cols::NTuple{N,Symbol} where N) = d[cols]
+    _filter(d::T, rows) where T<:Data.Table = T([col[rows] for col in d]...)
+end
+
+_size(d::Data.Table) = size(Data.schema(d))
+_size(d::Data.Table, dim::Int) = size(Data.schema(d), dim)
+
 ## Default NULL handler.  Others can be added as keyword arguments
-function missing_omit(df::DataFrame)
-    cc = completecases(df)
-    df[cc,:], cc
+function missing_omit(d::T) where T<:Data.Table
+    nonmissings = trues(_size(d, 1))
+    for col in d
+        _nonmissing!(nonmissings, col)
+    end
+    _filter(d, nonmissings), nonmissings
 end
 
 _droplevels!(x::Any) = x
 _droplevels!(x::AbstractCategoricalArray) = droplevels!(x)
 
-function ModelFrame(trms::Terms, d::AbstractDataFrame;
+function ModelFrame(trms::Terms, d::Data.Table;
                     contrasts::Dict = Dict())
-    df, msng = missing_omit(DataFrame(map(x -> d[x], trms.eterms), Symbol.(trms.eterms)))
-    names!(df, Symbol.(string.(trms.eterms)))
+    d, msng = missing_omit(_select(d, (Symbol.(trms.eterms)...)))
 
-    evaledContrasts = evalcontrasts(df, contrasts)
+    evaledContrasts = evalcontrasts(d, contrasts)
 
     ## Check for non-redundant terms, modifying terms in place
-    check_non_redundancy!(trms, df)
+    check_non_redundancy!(trms, d)
 
-    ModelFrame(df, trms, msng, evaledContrasts)
+    ModelFrame(d, trms, msng, evaledContrasts)
 end
 
-ModelFrame(df::AbstractDataFrame, term::Terms, msng::BitArray) = ModelFrame(df, term, msng, evalcontrasts(df))
-ModelFrame(f::Formula, d::AbstractDataFrame; kwargs...) = ModelFrame(Terms(f), d; kwargs...)
-ModelFrame(ex::Expr, d::AbstractDataFrame; kwargs...) = ModelFrame(Formula(ex), d; kwargs...)
+ModelFrame(f::Formula, d; kwargs...) = ModelFrame(Terms(f), d; kwargs...)
```

Comment:
Any reason not to mark `d::Data.Table`? In general I prefer being explicit about what types are accepted, it makes errors clearer and helps documenting the function.

Reply:
All of these constructors end up at one that does specify `Data.Table`.  My idea was to accept anything and let DataStreams sort out whether it could be streamed to a `Table`.  There's not a type restriction that specifies that, so we have to take Any.  The alternative would be to restrict the `ModelFrame` constructors to `Data.Table`, but allow anything in the `fit` methods.  I'd prefer this way to that, because then package authors who are constructing model frames manually also get the benefits of full DataStreams support without having to know about it.